drivers/sensor/pcnt_esp32/pcnt_esp32.c

```diff
@@ -0,0 +1,415 @@
+/*
+ * Copyright (c) 2022 Espressif Systems (Shanghai) Co., Ltd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT espressif_esp32_pcnt
+
+/* Include esp-idf headers first to avoid redefining BIT() macro */
+#include <hal/pcnt_hal.h>
+#include <hal/pcnt_ll.h>
+#include <hal/pcnt_types.h>
+
+#include <soc.h>
+#include <errno.h>
+#include <string.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/clock_control.h>
+#ifdef CONFIG_PCNT_ESP32_TRIGGER
+#include <zephyr/drivers/interrupt_controller/intc_esp32.h>
+#endif /* CONFIG_PCNT_ESP32_TRIGGER */
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(pcnt_esp32, CONFIG_SENSOR_LOG_LEVEL);
+
+#define PCNT_INTR_UNIT_0  BIT(0)
+#define PCNT_INTR_UNIT_1  BIT(1)
+#define PCNT_INTR_UNIT_2  BIT(2)
+#define PCNT_INTR_UNIT_3  BIT(3)
+#ifdef CONFIG_SOC_ESP32
+#define PCNT_INTR_UNIT_4  BIT(4)
+#define PCNT_INTR_UNIT_5  BIT(5)
+#define PCNT_INTR_UNIT_6  BIT(6)
+#define PCNT_INTR_UNIT_7  BIT(7)
+#endif /* CONFIG_SOC_ESP32 */
+
+#ifdef CONFIG_PCNT_ESP32_TRIGGER
+#define PCNT_INTR_THRES_1 BIT(2)
+#define PCNT_INTR_THRES_0 BIT(3)
+#endif /* CONFIG_PCNT_ESP32_TRIGGER */
+
+struct pcnt_esp32_data {
+	pcnt_hal_context_t hal;
+	struct k_sem cmd_sem;
+#ifdef CONFIG_PCNT_ESP32_TRIGGER
+	sensor_trigger_handler_t trigger_handler;
+#endif /* CONFIG_PCNT_ESP32_TRIGGER */
+};
+
+struct pcnt_esp32_channel_config {
+	uint8_t sig_pos_mode;
+	uint8_t sig_neg_mode;
+	uint8_t ctrl_h_mode;
+	uint8_t ctrl_l_mode;
+};
+
+struct pcnt_esp32_unit_config {
+	const uint8_t idx;
+	int16_t filter;
+	int16_t count_val_acc;
+	struct pcnt_esp32_channel_config channel_config[2];
+	int32_t h_thr;
+	int32_t l_thr;
+	int32_t offset;
+};
+
+struct pcnt_esp32_config {
+	const struct pinctrl_dev_config *pincfg;
+	const struct device *clock_dev;
+	const clock_control_subsys_t clock_subsys;
+	struct pcnt_esp32_unit_config *unit_config;
+	const int unit_len;
+};
+
+static int pcnt_esp32_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	const struct pcnt_esp32_config *config = dev->config;
+	struct pcnt_esp32_data *data = (struct pcnt_esp32_data *const)(dev)->data;
+
+	if (chan != SENSOR_CHAN_ALL && chan != SENSOR_CHAN_ROTATION) {
+		return -ENOTSUP;
+	}
+
+	k_sem_take(&data->cmd_sem, K_FOREVER);
```

Comment:
cmd_sem should not be a mutex instead of a plain semaphore? I'm afraid of cases of priority inversion in the block between semaphores.

Reply:
Thanks! I will change it to a mutex